# Machine Learning Engineer Nanodegree
## Supervised Learning
## Project 2: Building a Student Intervention System

Welcome to the second project of the Machine Learning Engineer Nanodegree! In this notebook, some template code has already been provided for you, and it will be your job to implement the additional functionality necessary to successfully complete this project. Sections that begin with **'Implementation'** in the header indicate that the following block of code will require additional functionality which you must provide. Instructions will be provided for each section and the specifics of the implementation are marked in the code block with a `'TODO'` statement. Please be sure to read the instructions carefully!

In addition to implementing code, there will be questions that you must answer which relate to the project and your implementation. Each section where you will answer a question is preceded by a **'Question X'** header. Carefully read each question and provide thorough answers in the following text boxes that begin with **'Answer:'**. Your project submission will be evaluated based on your answers to each of the questions and the implementation you provide.  

>**Note:** Code and Markdown cells can be executed using the **Shift + Enter** keyboard shortcut. In addition, Markdown cells can be edited by typically double-clicking the cell to enter edit mode.

### Question 1 - Classification vs. Regression
*Your goal for this project is to identify students who might need early intervention before they fail to graduate. Which type of supervised learning problem is this, classification or regression? Why?*

**Answer: ** Classification - we are filtering the whole group into two categories - “Needs Intervention” and “Doesn’t Need Intervention”. Regression is more suited for situations where one would be predicting continuous values. In our situation, we are predicting which group each student belongs to.

## Exploring the Data
Run the code cell below to load necessary Python libraries and load the student data. Note that the last column from this dataset, `'passed'`, will be our target label (whether the student graduated or didn't graduate). All other columns are features about each student.

In [7]:
# Import libraries
import numpy as np
import pandas as pd
from time import time
from sklearn.metrics import f1_score

# Read student data
student_data = pd.read_csv("student-data.csv")
print "Student data read successfully!"

Student data read successfully!


### Implementation: Data Exploration
Let's begin by investigating the dataset to determine how many students we have information on, and learn about the graduation rate among these students. In the code cell below, you will need to compute the following:
- The total number of students, `n_students`.
- The total number of features for each student, `n_features`.
- The number of those students who passed, `n_passed`.
- The number of those students who failed, `n_failed`.
- The graduation rate of the class, `grad_rate`, in percent (%).


In [8]:
# TODO: Calculate number of students
n_students = student_data.shape[0] # another method would be len(student_data)

# TODO: Calculate number of features
n_features = student_data.shape[1] - 1 #subtract 1 to compensate for labels

# TODO: Calculate passing students
n_passed = student_data[student_data['passed']=='yes'].shape[0]

# TODO: Calculate failing students
n_failed = student_data[student_data['passed']=='no'].shape[0] #n_failed = n_students - n_passed also works

# TODO: Calculate graduation rate
grad_rate = (n_passed * 100.00) / n_students

# Print the results
print "Total number of students: {}".format(n_students)
print "Number of features: {}".format(n_features)
print "Number of students who passed: {}".format(n_passed)
print "Number of students who failed: {}".format(n_failed)
print "Graduation rate of the class: {:.2f}%".format(grad_rate)

Total number of students: 395
Number of features: 30
Number of students who passed: 265
Number of students who failed: 130
Graduation rate of the class: 67.09%


## Preparing the Data
In this section, we will prepare the data for modeling, training and testing.

### Identify feature and target columns
It is often the case that the data you obtain contains non-numeric features. This can be a problem, as most machine learning algorithms expect numeric data to perform computations with.

Run the code cell below to separate the student data into feature and target columns to see if any features are non-numeric.

In [9]:
# Extract feature columns
feature_cols = list(student_data.columns[:-1])

# Extract target column 'passed'
target_col = student_data.columns[-1] 

# Show the list of columns
print "Feature columns:\n{}".format(feature_cols)
print "\nTarget column: {}".format(target_col)

# Separate the data into feature data and target data (X_all and y_all, respectively)
X_all = student_data[feature_cols]
y_all = student_data[target_col]

# Show the feature information by printing the first five rows
print "\nFeature values:"
print X_all.head()

Feature columns:
['school', 'sex', 'age', 'address', 'famsize', 'Pstatus', 'Medu', 'Fedu', 'Mjob', 'Fjob', 'reason', 'guardian', 'traveltime', 'studytime', 'failures', 'schoolsup', 'famsup', 'paid', 'activities', 'nursery', 'higher', 'internet', 'romantic', 'famrel', 'freetime', 'goout', 'Dalc', 'Walc', 'health', 'absences']

Target column: passed

Feature values:
  school sex  age address famsize Pstatus  Medu  Fedu     Mjob      Fjob  \
0     GP   F   18       U     GT3       A     4     4  at_home   teacher   
1     GP   F   17       U     GT3       T     1     1  at_home     other   
2     GP   F   15       U     LE3       T     1     1  at_home     other   
3     GP   F   15       U     GT3       T     4     2   health  services   
4     GP   F   16       U     GT3       T     3     3    other     other   

    ...    higher internet  romantic  famrel  freetime goout Dalc Walc health  \
0   ...       yes       no        no       4         3     4    1    1      3   
1   ...       

### Preprocess Feature Columns

As you can see, there are several non-numeric columns that need to be converted! Many of them are simply `yes`/`no`, e.g. `internet`. These can be reasonably converted into `1`/`0` (binary) values.

Other columns, like `Mjob` and `Fjob`, have more than two values, and are known as _categorical variables_. The recommended way to handle such a column is to create as many columns as possible values (e.g. `Fjob_teacher`, `Fjob_other`, `Fjob_services`, etc.), and assign a `1` to one of them and `0` to all others.

These generated columns are sometimes called _dummy variables_, and we will use the [`pandas.get_dummies()`](http://pandas.pydata.org/pandas-docs/stable/generated/pandas.get_dummies.html?highlight=get_dummies#pandas.get_dummies) function to perform this transformation. Run the code cell below to perform the preprocessing routine discussed in this section.

In [10]:
def preprocess_features(X):
    ''' Preprocesses the student data and converts non-numeric binary variables into
        binary (0/1) variables. Converts categorical variables into dummy variables. '''
    
    # Initialize new output DataFrame
    output = pd.DataFrame(index = X.index)

    # Investigate each feature column for the data
    for col, col_data in X.iteritems():
        
        # If data type is non-numeric, replace all yes/no values with 1/0
        if col_data.dtype == object:
            col_data = col_data.replace(['yes', 'no'], [1, 0])

        # If data type is categorical, convert to dummy variables
        if col_data.dtype == object:
            # Example: 'school' => 'school_GP' and 'school_MS'
            col_data = pd.get_dummies(col_data, prefix = col)  
        
        # Collect the revised columns
        output = output.join(col_data)
    
    return output

X_all = preprocess_features(X_all)
print "Processed feature columns ({} total features):\n{}".format(len(X_all.columns), list(X_all.columns))

Processed feature columns (48 total features):
['school_GP', 'school_MS', 'sex_F', 'sex_M', 'age', 'address_R', 'address_U', 'famsize_GT3', 'famsize_LE3', 'Pstatus_A', 'Pstatus_T', 'Medu', 'Fedu', 'Mjob_at_home', 'Mjob_health', 'Mjob_other', 'Mjob_services', 'Mjob_teacher', 'Fjob_at_home', 'Fjob_health', 'Fjob_other', 'Fjob_services', 'Fjob_teacher', 'reason_course', 'reason_home', 'reason_other', 'reason_reputation', 'guardian_father', 'guardian_mother', 'guardian_other', 'traveltime', 'studytime', 'failures', 'schoolsup', 'famsup', 'paid', 'activities', 'nursery', 'higher', 'internet', 'romantic', 'famrel', 'freetime', 'goout', 'Dalc', 'Walc', 'health', 'absences']


### Implementation: Training and Testing Data Split
So far, we have converted all _categorical_ features into numeric values. For the next step, we split the data (both features and corresponding labels) into training and test sets. In the following code cell below, you will need to implement the following:
- Randomly shuffle and split the data (`X_all`, `y_all`) into training and testing subsets.
  - Use 300 training points (approximately 75%) and 95 testing points (approximately 25%).
  - Set a `random_state` for the function(s) you use, if provided.
  - Store the results in `X_train`, `X_test`, `y_train`, and `y_test`.

In [13]:
# TODO: Import any additional functionality you may need here
from sklearn.cross_validation import train_test_split 
#future reference: Stratified Shuffle Split is preferred unbalanced data because it preserves the % of samples for each class
# TODO: Set the number of training points
num_train = 300

# Set the number of testing points
num_test = X_all.shape[0] - num_train

# TODO: Shuffle and split the dataset into the number of training and testing points above
X_train, X_test, y_train, y_test = train_test_split(X_all, y_all, test_size=num_test, train_size=num_train, random_state=125)

# Show the results of the split
print "Training set has {} samples.".format(X_train.shape[0])
print "Testing set has {} samples.".format(X_test.shape[0])

Training set has 300 samples.
Testing set has 95 samples.


## Training and Evaluating Models
In this section, you will choose 3 supervised learning models that are appropriate for this problem and available in `scikit-learn`. You will first discuss the reasoning behind choosing these three models by considering what you know about the data and each model's strengths and weaknesses. You will then fit the model to varying sizes of training data (100 data points, 200 data points, and 300 data points) and measure the F<sub>1</sub> score. You will need to produce three tables (one for each model) that shows the training set size, training time, prediction time, F<sub>1</sub> score on the training set, and F<sub>1</sub> score on the testing set.

**The following supervised learning models are currently available in** [`scikit-learn`](http://scikit-learn.org/stable/supervised_learning.html) **that you may choose from:**
- Gaussian Naive Bayes (GaussianNB)
- Decision Trees
- Ensemble Methods (Bagging, AdaBoost, Random Forest, Gradient Boosting)
- K-Nearest Neighbors (KNeighbors)
- Stochastic Gradient Descent (SGDC)
- Support Vector Machines (SVM)
- Logistic Regression

### Question 2 - Model Application
*List three supervised learning models that are appropriate for this problem. For each model chosen*
- Describe one real-world application in industry where the model can be applied. *(You may need to do a small bit of research for this — give references!)* 
- What are the strengths of the model; when does it perform well? 
- What are the weaknesses of the model; when does it perform poorly?
- What makes this model a good candidate for the problem, given what you know about the data?

# **Answer: ** 
Model 1: Decision Trees 

Real World Application: 
Choosing a restaurant for dinner. Let's say you are trying to pick the ideal restaurant for a hot dinner date. To make your decision, you may consider a number of different factors about the restaurant such as cost, type of food, waiting time, distance, ambiance, etc. The Decision Tree method will allow you to classify a particular restaurant as 'suitable' or 'unsuitable' based off of your predefined preferences. You've got too much else to worry about for your date! Alleviate your decision with a Decision Tree.

Strengths: 
Intuitive and easy to understand. Requires minimal data prep (fitting, normalizing, etc.). Can handle numerical & categorical data. Good for discrete and continuous data. Can be validated with statistical tests. Good with large datasets and robust enough to satisfy tons of different applications.

Weaknesses: 
Decision Trees have the tendency to be greedy - i.e. they fail to reconsider their input or think about future inputs. Can be prone to overfitting when the number of features increases. Can become prehibitively large when used in conjuction with various tests. Can be biased toward attributes with more levels.

Why is it a good candidate? 
Our dataset contains a large portion of features that are categorical or binary, and most of the features have a small number of possible outcomes. Setting up our decision tree should be much more intuitive than it would be if our feature outcomes were not as uniform or simple. However, we are dealing with a relatively small dataset that is rich in many, varying features which could lead to overfitting.


Model 2: Support Vector Machines (SVM)

Real World Application: 
Predicting a person's gender based on information on height and hair-length. Say you have a dataset describing a population of males and females. Using SVM, you are able to develop a set of rules that can be used to classify each person as either Male or Female. If a person is taller than average and has shorter-than-average hair, signs would point to that person being a male. And vise versa for female: shorter height and longer hair. SVMs are also commonly used in image recognition, speech recognition, and face detection.

Strengths: 
SVMs are effective in high-dimensional spaces, even in situations where there are more features than total samples! SVMs are versatile because one can use kernels to further drive data toward linear separability. They are also memory efficient, which is never a bad thing!

Weaknesses: 
SVMs don't do well with large datasets. They struggle with noisy overlapping data. And they do not provide probability estimates for their outputs (i.e. person is 60% likely to be male / 40% female - SVMs will classify as one or the other, but won't provide any in-between). 

Why is it a good candidate? 
We have a situation where there are relatively few data sets and a vast number of dimensions. SVM is definitely a suitable option in this scenario.

Model 3: Random Forest

Real World Application: 
Planning a picnic in the park with friends and you want to predict if the weather will be nice or not. Your factors of consideration may be the weather forecast: sunny, rainy, or overcast. Other factors may also be considered: humidity level, wind speeds, etc. Random Forest compares many random subsets of data considering a number of random predictor variables. The result is many different trees that can shed insight into the interaction between a combination of predictor variables.

Strengths: 
Random Forests can deal with unbalanced and missing data. Most importantly, they can alleviate over-fitting limitations imposed by individual learning trees by growing multiple distinct decision trees that use varying samples and focus on different subsets of features.

Weaknesses: 
Accuracy of predictions are limited by the range of training data. Random Forests may be overfitting in datasets that are particularly noisy. Because Random Forests rely on resampling and fitting many decision trees instead of one, there may be challenges with computational efficiency compared to other options.

Why is it a good candidate?
Random Forest is a helpful technique in dealing with the over-fitting tendencies of a single decision tree. Like Decision Trees, Random Forests also do well with features consisting of mixed data types, making it a suitable ensemble method to use in addition to Decision Trees.



Sources:

Berwick, R. "Support Vector Machines." SpringerReference (n.d.): n. pag. Web. 4 Sept. 2016. <http://web.mit.edu/6.034/wwwbob/svm-notes-long-08.pdf>.

Bhalla, Deepanshu. "Listen Data." Random Forest Explained in Simple Terms -. N.p., n.d. Web. 04 Sept. 2016. <http://www.listendata.com/2014/11/random-forest-with-r.html>.

By Contrast, in a Black Box Model (e.g., in an Artificial Neural. "1.10. Decision Trees¶." 1.10. Decision Trees — Scikit-learn 0.17.1 Documentation. N.p., n.d. Web. 4 Sept. 2016.

"Decision Trees." SpringerReference (n.d.): n. pag. Web. 4 Sept. 2016.
Https://www.analyticsvidhya.com/blog/author/sunil-ray/. "Understanding Support Vector Machine 

Algorithm from Examples (along with Code)." Analytics Vidhya. N.p., 11 Dec. 2015. Web. 04 Sept. 2016.

### Setup
Run the code cell below to initialize three helper functions which you can use for training and testing the three supervised learning models you've chosen above. The functions are as follows:
- `train_classifier` - takes as input a classifier and training data and fits the classifier to the data.
- `predict_labels` - takes as input a fit classifier, features, and a target labeling and makes predictions using the F<sub>1</sub> score.
- `train_predict` - takes as input a classifier, and the training and testing data, and performs `train_clasifier` and `predict_labels`.
 - This function will report the F<sub>1</sub> score for both the training and testing data separately.

In [35]:
def train_classifier(clf, X_train, y_train):
    ''' Fits a classifier to the training data. '''
    
    # Start the clock, train the classifier, then stop the clock
    start = time()
    clf.fit(X_train, y_train)
    end = time()
    
    # Print the results
    print "Trained model in {:.4f} seconds".format(end - start)

    
def predict_labels(clf, features, target):
    ''' Makes predictions using a fit classifier based on F1 score. '''
    
    # Start the clock, make predictions, then stop the clock
    start = time()
    y_pred = clf.predict(features)
    end = time()
    
    # Print and return results
    print "Made predictions in {:.4f} seconds.".format(end - start)
    return f1_score(target.values, y_pred, pos_label='yes')


def train_predict(clf, X_train, y_train, X_test, y_test):
    ''' Train and predict using a classifer based on F1 score. '''
    
    # Indicate the classifier and the training set size
    print "Training a {} using a training set size of {}. . .".format(clf.__class__.__name__, len(X_train))
    
    # Train the classifier
    train_classifier(clf, X_train, y_train)
    
    # Print the results of prediction for both training and testing
    print "F1 score for training set: {:.4f}.".format(predict_labels(clf, X_train, y_train))
    print "F1 score for test set: {:.4f}.".format(predict_labels(clf, X_test, y_test))

### Implementation: Model Performance Metrics
With the predefined functions above, you will now import the three supervised learning models of your choice and run the `train_predict` function for each one. Remember that you will need to train and predict on each classifier for three different training set sizes: 100, 200, and 300. Hence, you should expect to have 9 different outputs below — 3 for each model using the varying training set sizes. In the following code cell, you will need to implement the following:
- Import the three supervised learning models you've discussed in the previous section.
- Initialize the three models and store them in `clf_A`, `clf_B`, and `clf_C`.
 - Use a `random_state` for each model you use, if provided.
 - **Note:** Use the default settings for each model — you will tune one specific model in a later section.
- Create the different training set sizes to be used to train each model.
 - *Do not reshuffle and resplit the data! The new training points should be drawn from `X_train` and `y_train`.*
- Fit each model with each training set size and make predictions on the test set (9 in total).  
**Note:** Three tables are provided after the following code cell which can be used to store your results.

In [25]:
# TODO: Import the three supervised learning models from sklearn
from sklearn.tree import DecisionTreeClassifier
from sklearn.svm import SVC
from sklearn.ensemble import RandomForestClassifier

# TODO: Initialize the three models
clf_A = DecisionTreeClassifier(random_state = 35)
clf_B = SVC(random_state = 35)
clf_C = RandomForestClassifier(random_state = 35)

# TODO: Set up the training set sizes
X_train_100 = X_train[:100]
y_train_100 = y_train[:100]

X_train_200 = X_train[:200]
y_train_200 = y_train[:200]

X_train_300 = X_train[:300]
y_train_300 = y_train[:300]

for clf in [clf_A, clf_B, clf_C]:
    #future reference: use a for loop on code below instead of typing it all out
    for train in [(X_train_100, y_train_100), (X_train_200, y_train_200), (X_train_300, y_train_300)]:
        train_predict(clf, train[0], train[1], X_test, y_test)
        print

# TODO: Execute the 'train_predict' function for each classifier and each training set size
# train_predict(clf, X_train, y_train, X_test, y_test)

Training a DecisionTreeClassifier using a training set size of 100. . .
Trained model in 0.0020 seconds
Made predictions in 0.0003 seconds.
F1 score for training set: 1.0000.
Made predictions in 0.0003 seconds.
F1 score for test set: 0.6767.

Training a DecisionTreeClassifier using a training set size of 200. . .
Trained model in 0.0020 seconds
Made predictions in 0.0003 seconds.
F1 score for training set: 1.0000.
Made predictions in 0.0002 seconds.
F1 score for test set: 0.7259.

Training a DecisionTreeClassifier using a training set size of 300. . .
Trained model in 0.0024 seconds
Made predictions in 0.0009 seconds.
F1 score for training set: 1.0000.
Made predictions in 0.0010 seconds.
F1 score for test set: 0.7194.

Training a SVC using a training set size of 100. . .
Trained model in 0.0024 seconds
Made predictions in 0.0017 seconds.
F1 score for training set: 0.9155.
Made predictions in 0.0015 seconds.
F1 score for test set: 0.7639.

Training a SVC using a training set size of 200

### Tabular Results
Edit the cell below to see how a table can be designed in [Markdown](https://github.com/adam-p/markdown-here/wiki/Markdown-Cheatsheet#tables). You can record your results from above in the tables provided.

** Classifer 1 - Decision Tree**  

| Training Set Size | Training Time (s) | Prediction Time (test) | F1 Score (train) | F1 Score (test) |
| :---------------: | :---------------------: | :--------------------: | :--------------: | :-------------: |
| 100 | 0.0020    |   0.0003   |  1.0000     |  0.6767    |
| 200 | 0.0020    |   0.0002   |  1.0000     |  0.7259    |
| 300 | 0.0024    |   0.0010   |  1.0000     |  0.7194    |

** Classifer 2 - SVM**  

| Training Set Size | Training Time (s) | Prediction Time (test) | F1 Score (train) | F1 Score (test) |
| :---------------: | :---------------------: | :--------------------: | :--------------: | :-------------: |
| 100 | 0.0024    |   0.0015   |  0.9155    |  0.7639    |
| 200 | 0.0056    |   0.0024   |  0.8817     |  0.7429    |
| 300 | 0.0110    |   0.0032   |  0.8770     |  0.7895    |

** Classifer 3 - Random Forest**  

| Training Set Size | Training Time (s) | Prediction Time (test) | F1 Score (train) | F1 Score (test) |
| :---------------: | :---------------------: | :--------------------: | :--------------: | :-------------: |
| 100 | 0.0461    |   0.0012   |  0.9922     |  0.7246    |
| 200 | 0.0351    |   0.0014   |  0.9921     |  0.7500    |
| 300 | 0.0288    |   0.0013   |  1.0000     |  0.7352    |

## Choosing the Best Model
In this final section, you will choose from the three supervised learning models the *best* model to use on the student data. You will then perform a grid search optimization for the model over the entire training set (`X_train` and `y_train`) by tuning at least one parameter to improve upon the untuned model's F<sub>1</sub> score. 

### Question 3 - Choosing the Best Model
*Based on the experiments you performed earlier, in one to two paragraphs, explain to the board of supervisors what single model you chose as the best model. Which model is generally the most appropriate based on the available data, limited resources, cost, and performance?*

**Answer: ** Of the three models used, I believe the Support Vector Machines (SVM) model is the most appropriate considering our data, cost, resource and performance constraints. It yeilded a higher F1 score in comparison to the other models, suggesting it most accurately identifies and appropriately classifies at-risk students in need of an intervention (recall and precision). Training and testing times of the SVM model were relatively efficient compared to the other models, and much faster than the Random Forest model. In addition to having the highest F1 score, the SVM model does not have as high of a risk of over-fitting the data as the Decision Tree model, and the SVM model is suitable to be used in conjunction with additional tests (kernels) that can further improve the model's confidence level.

### Question 4 - Model in Layman's Terms
*In one to two paragraphs, explain to the board of directors in layman's terms how the final model chosen is supposed to work. Be sure that you are describing the major qualities of the model, such as how the model is trained and how the model makes a prediction. Avoid using advanced mathematical or technical jargon, such as describing equations or discussing the algorithm implementation.*

**Answer: ** The goal for this project was to accurately identify students who might need an early intervention before they fail to graduate. We started by gathering all possible information we had on 395 students, consisting of information such as: age, whether the student has internet access at home, number of absences, and more (30 different features in total). Of those 395 students, 265 students passed and 130 failed. Because we have a vast amount of features for each student but a relatively small sample size of students to analyze, it was deemed that a Support Vector Machine model (SVM) would be a suitable candidate for analysis. 

The Support Vector Machine model works to identify commonly found behaviors or traits that differ between students that graduated and those that failed. It does this by establishing a decision boundary that acts as a border between the two groups. An optimal decision boundary is one that, not only, draws a dividing line between the two groups, but also maximizes the margins of separation of the groups. The illustration below should provide a valuable visualisation. A good example would be to the Korean Demilitarized Zone that separates North and South Korea. Imagine the "Best Line" as the border itself, but each "Support Line" acts as an additional barrier that prevents individuals from getting too close to the border. By maximizing the distance of separation, we should have a clearer distinction between the two groups.
<img src="https://udacity-github-sync-content.s3.amazonaws.com/_imgs/8650/1461641603/SVM_2.png">

One issue arises when we are faced with a nonlinearly-separable dataset (try saying that three times fast!) where there isn't a clear visual separation between the two groups. Fortunately for us, the SVM model has a few tricks up its sleeve! By using what's known as the 'Kernel Trick' (illustrated below), we are able to tweak the unseparable data points into a more digestable form that ALSO maintains the core relationships between the points. We're essentially translating the data from a foreign language into our native tongue, where the story stays the same, but is told in a language we can better understand.
<img src="https://udacity-github-sync-content.s3.amazonaws.com/_imgs/8650/1461641606/SVM.png">

When faced with information on a new student, the model can now chart where he/she fits in relation to the dividing line, and predict which group the new student belongs with. With a certain level of confidence, we now have a better idea of what types of behaviors should start to raise red flags that we can address before a student ends up on a wayward path.

### Implementation: Model Tuning
Fine tune the chosen model. Use grid search (`GridSearchCV`) with at least one important parameter tuned with at least 3 different values. You will need to use the entire training set for this. In the code cell below, you will need to implement the following:
- Import [`sklearn.grid_search.gridSearchCV`](http://scikit-learn.org/stable/modules/generated/sklearn.grid_search.GridSearchCV.html) and [`sklearn.metrics.make_scorer`](http://scikit-learn.org/stable/modules/generated/sklearn.metrics.make_scorer.html).
- Create a dictionary of parameters you wish to tune for the chosen model.
 - Example: `parameters = {'parameter' : [list of values]}`.
- Initialize the classifier you've chosen and store it in `clf`.
- Create the F<sub>1</sub> scoring function using `make_scorer` and store it in `f1_scorer`.
 - Set the `pos_label` parameter to the correct value!
- Perform grid search on the classifier `clf` using `f1_scorer` as the scoring method, and store it in `grid_obj`.
- Fit the grid search object to the training data (`X_train`, `y_train`), and store it in `grid_obj`.

In [43]:
# TODO: Import 'GridSearchCV' and 'make_scorer'
from sklearn.grid_search import GridSearchCV
from sklearn.metrics import make_scorer

# TODO: Create the parameters list you wish to tune
parameters = {"C": [0.01, 0.1, 1.0, 10.0, 100.0], "kernel": ['rbf', 'linear', 'poly', 'sigmoid']}

# TODO: Initialize the classifier
clf = SVC(random_state = 35)

# TODO: Make an f1 scoring function using 'make_scorer' 
f1_scorer = make_scorer(f1_score, pos_label="yes")

# TODO: Perform grid search on the classifier using the f1_scorer as the scoring method
grid_obj = GridSearchCV(clf, parameters, scoring = f1_scorer)

# TODO: Fit the grid search object to the training data and find the optimal parameters
grid_obj = grid_obj.fit(X_train, y_train)

# Get the estimator
clf = grid_obj.best_estimator_

# Report the final F1 score for training and testing after parameter tuning
print "Tuned model has a training F1 score of {:.4f}.".format(predict_labels(clf, X_train, y_train))
print "Tuned model has a testing F1 score of {:.4f}.".format(predict_labels(clf, X_test, y_test))

Made predictions in 0.0036 seconds.
Tuned model has a training F1 score of 0.8493.
Made predictions in 0.0014 seconds.
Tuned model has a testing F1 score of 0.7755.


### Question 5 - Final F<sub>1</sub> Score
*What is the final model's F<sub>1</sub> score for training and testing? How does that score compare to the untuned model?*

**Answer: ** The final F1 score for training is .8493, and the F1 score for testing is 0.7755. Let's compare to the untuned model with a training set of 200. The untuned model had training F1 and testing F1 scores of 0.8817 and 0.7429, respectively. The training F1 score went down slightly, but the testing F1 improved. Because there is not a significant improvement between the tuned and untuned models, one might conclude that the default C and kernel values were initially appropriate. By continuing to adjust the tuning parameters, we should find a set of parameters that will raise our level of confidence and reduce the tendency to overfit.

> **Note**: Once you have completed all of the code implementations and successfully answered each question above, you may finalize your work by exporting the iPython Notebook as an HTML document. You can do this by using the menu above and navigating to  
**File -> Download as -> HTML (.html)**. Include the finished document along with this notebook as your submission.